sql/core/src/main/scala/org/apache/spark/sql/execution/dynamicpruning/PartitionPruning.scala

```diff
@@ -183,28 +184,20 @@ object PartitionPruning extends Rule[LogicalPlan] with PredicateHelper with Join
     })
     val scanOverhead = notCached.map(_.stats.sizeInBytes).sum.toFloat
     val cachedOverhead = cached.map {
-      case m: InMemoryRelation if m.cacheBuilder.storageLevel.useDisk &&
-          !m.cacheBuilder.storageLevel.useMemory =>
-        m.stats.sizeInBytes.toFloat
-      case m: InMemoryRelation if m.cacheBuilder.storageLevel.useDisk =>
-        m.stats.sizeInBytes.toFloat * 0.2
-      case m: InMemoryRelation if m.cacheBuilder.storageLevel.useMemory =>
-        0.0
+      case m: InMemoryRelation if m.isMaterialized =>
+        if (m.cacheBuilder.storageLevel.useDisk && !m.cacheBuilder.storageLevel.useMemory) {
+          m.stats.sizeInBytes.toFloat
+        } else if (m.cacheBuilder.storageLevel.useDisk) {
+          m.stats.sizeInBytes.toFloat * 0.2
+        } else {
+          // use memory only
+          0.0
+        }
+      case m => m.stats.sizeInBytes.toFloat
```

Comment:
Hmm, this looks okay, although I guess that the cost of materializing maybe be still larger for `useDisk` (writing to disk?) than `useMemory` too?

Reply:
yeas, to be conservative I use the overhead of original plan, otherwise it may cause different behavior with same query but one is cached